### PR TITLE
Fix URL handling in the whole markdown module, improve test coverage (fix #997)

### LINKED ISSUE #997: Local instance of gitea don`t display images.
- Can you reproduce the bug at https://try.gitea.io:
  - [X] Yes (https://try.gitea.io/Lourens-Rich/markdown-image-bug/src/master/2017/02/note.md)
  - [  ] No.
  - [  ] Not relevant

On this link (https://git.r208.tk/Lourens-Rich/markdown-image-bug/src/master/2017/02/note.md) Gitea don't display image. (All repos images on my instance now are broken.)

### PR DESCRIPTION
**Pull request by andrew-boyarshin**
As it is in the title.

Plus, links to issues or PRs in external repositories will no longer be rewritten to current repo.

Fixes issue #997. Merging this PR will increase coverage to 32.4% from 31.8%.

### CODE REVIEW

File: vendor/code.gitea.io/sdk/gitea/miscellaneous.go [hunk 1]
```diff
@@ -9,4 +9,5 @@ type MarkdownOption struct {
 	Text    string
 	Mode    string
 	Context string
+	Wiki    string
```
Comment (andrew-boyarshin) [inline vendor/code.gitea.io/sdk/gitea/miscellaneous.go]: Public Gitea API changed, optional Wiki parameter added.

File: routers/api/v1/misc/markdown_test.go [hunk 2]
```diff
@@ -74,7 +75,7 @@ func TestAPI_RenderGFM(t *testing.T) {
 <ul>
 <li><a href="` + AppSubURL + `wiki/Links" rel="nofollow">Links, Language bindings, Engine bindings</a></li>
 <li><a href="` + AppSubURL + `wiki/Tips" rel="nofollow">Tips</a></li>
-<li>Bezier widget (by <a href="` + AppURL + `r-lyeh" rel="nofollow">@r-lyeh</a>)<a href="` + AppSubURL + `issues/786" rel="nofollow">#786</a></li>
+<li>Bezier widget (by <a href="` + AppURL + `r-lyeh" rel="nofollow">@r-lyeh</a>)<a href="https://github.com/ocornut/imgui/issues/786" rel="nofollow">#786</a></li>
```
Comment (andrew-boyarshin) [inline routers/api/v1/misc/markdown_test.go:78]: Note, link hosts are no longer rewritten. No more dead links.

File: modules/markdown/markdown.go [hunk 3]
```diff
@@ -246,10 +245,30 @@ func URLJoin(elem ...string) string {
 	last := len(elem) - 1
 	for i, item := range elem {
 		res += item
-		if !strings.HasSuffix(res, "/") && i != last {
+		if i != last && !strings.HasSuffix(res, "/") {
 			res += "/"
 		}
 	}
+	cwdIndex := strings.Index(res, "/./")
```
Comment (andrew-boyarshin) [inline modules/markdown/markdown.go:252]: ```URLJoin``` now correctly handles ```.``` and ```..``` path elements.

File: modules/markdown/markdown.go [hunk 4]
```diff
@@ -92,10 +92,10 @@ var (
 	ShortLinkPattern = regexp.MustCompile(`(\[\[.*\]\]\w*)`)
 
 	// AnySHA1Pattern allows to split url containing SHA into parts
-	AnySHA1Pattern = regexp.MustCompile(`http\S+//(\S+)/(\S+)/(\S+)/(\S+)/([0-9a-f]{40})(?:/?([^#\s]+)?(?:#(\S+))?)?`)
+	AnySHA1Pattern = regexp.MustCompile(`(http\S*)://(\S+)/(\S+)/(\S+)/(\S+)/([0-9a-f]{40})(?:/?([^#\s]+)?(?:#(\S+))?)?`)
```
Comment (andrew-boyarshin) [inline modules/markdown/markdown.go:95]: Protocol is important now.

File: vendor/code.gitea.io/sdk/gitea/miscellaneous.go (hunk 1 above)
Comment (lunny) [reply vendor/code.gitea.io/sdk/gitea/miscellaneous.go]: But you have to send a PR to SDK project at first before this can be merged.

**Review by lunny — CHANGES_REQUESTED** [review]

File: vendor/code.gitea.io/sdk/gitea/miscellaneous.go (hunk 1 above)
Comment (andrew-boyarshin) [reply vendor/code.gitea.io/sdk/gitea/miscellaneous.go]: @lunny done.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
LGTM except the Wiki type should be bool

**Comment by andrew-boyarshin** [comment]
In case someone hasn't noticed, I force-pushed amended commits to both gitea and gitea-sdk repos with string -> bool change. Review for the final LGTM?

**Comment by appleboy** [comment]
LGTM

**Review by appleboy — APPROVED** [review]

**Comment by andrew-boyarshin** [comment]
Pending merge.

**Pull request merged**